### PR TITLE
Added pushState fallback for IE 9 browser. fixes #6506

### LINKED ISSUE #6506: Angular 2, IE9 & pushState issue!
Hi,

I'm having some issues with Angular 2, pushState & IE9 as I've documented initially in https://groups.google.com/forum/#!topic/angular/TdCt6KyhwI8 before carrying out much more investigation!

I've read https://github.com/angular/angular/issues/5792 and attempted to apply the history.js script with mixed results!

I have had to modify angular2.dev.js locally to return window.History instead of window.history (not a long term solution really!)

In Everything but IE9 things work as expected using path location mode - in IE9 history.js takes over and injects urls to look like http://my.app.com/#./dashboard

But those hashtag paths are then not recognised by the angular router..!!

So I turn on the hash location strategy and that works fine in chrome etc but in IE history.js kicks in and alerts the following..

>> History.js does not support states with fragment-identifiers (hashes/anchors). 

The root issue is found in the following code in the router..

```
    PlatformLocation.prototype.pushState = function(state, title, url) {
        console.log('pushState line 2022');
      this._history.pushState(state, title, url);
      console.log('after push line 2026');
    };
```
If I comment out the pushState line then everything works.. but of course the url is not updated in the browser window :(

It feels a bit like a catch-22 situation!!

a) Is this a bug??
b) Does the Angular team have a compatible IE9 routing solution?

Help!!!

### PR DESCRIPTION
**Pull request by sasidhar**
- **The commit message follows our guidelines:**
  https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format
- **Tests for the changes have been added (for bug fixes / features)**
  No tests added, as the changes are only for IE9 browsers where `pushState` is not available
- **Docs have been added / updated (for bug fixes / features)**
  Not updated. As we are already mentioning that Angular 2 supports IE9, and as this is a bug fix, I don't see any need for updating Docs
- **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
  Only code changes
- **What is the current behavior?** (You can also link to an open issue here)
  #6506
- **What is the new behavior (if this is a feature change)?**
  This should work in IE9 without any `pushState` issues
- **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
  No

### CODE REVIEW

File: modules/angular2/src/router/location/browser_platform_location.ts [hunk 1]
```diff
@@ -45,11 +45,19 @@ export class BrowserPlatformLocation extends PlatformLocation {
   set pathname(newPath: string) { this._location.pathname = newPath; }
 
   pushState(state: any, title: string, url: string): void {
-    this._history.pushState(state, title, url);
+    if(this._history.pushState) {
```
Comment (gdi2290) [inline modules/angular2/src/router/location/browser_platform_location.ts]: can you import `import {isPresent} from 'angular2/src/facade/lang';` and use `isPresent(this._history.pushState)` here so dart will pass

File: modules/angular2/src/router/location/browser_platform_location.ts [hunk 2]
```diff
@@ -45,11 +45,19 @@ export class BrowserPlatformLocation extends PlatformLocation {
...
+      this._history.pushState(state, title, url);
+    } else if (this._location) {
+      this._location.hash = url;
+    }
   }
 
   replaceState(state: any, title: string, url: string): void {
-    this._history.replaceState(state, title, url);
+    if(this._history.replaceState) {
```
Comment (gdi2290) [inline modules/angular2/src/router/location/browser_platform_location.ts]: same for these, can you use `isPresent`

**Comment by sasidhar** [comment]
@gdi2290 Thanks. I will update the code and send another PR

**Comment by gdi2290** [comment]
can you rebase this current commit after you update it https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-submitting-a-pull-request-pr

**Comment by sasidhar** [comment]
@gdi2290 Sent another fresh PR with #7929 with updated code using `isPresent`

Closing this.

**Pull request closed without merging**